File: python/meetup/meetup.py
from datetime import date, timedelta
from calendar import monthrange

class MeetupDayException(BaseException):
    '''Raise when my specific value is wrong'''

def meetup_day(year, month, day, week):
	month_start = date(year, month, 1)
	first_day_with_name = __get_first_day_with_name(month_start, day)
	if week == 'teenth':
		week_start = date(year, month, 13)
		return __get_first_day_with_name(week_start, day)
	elif week == 'last':
		return __get_last_day_with_name(year, month, day)
	else:
		new_date = first_day_with_name + timedelta(days=delta_map[week])
		if new_date.month == month:
			return new_date
		else:
			raise MeetupDayException

def __get_first_day_with_name(start, day):
	day_number = day_map[day]
	start_day_number = start.weekday()
	offset = (day_number - start_day_number + 7) % 7
	return start + timedelta(days=offset)


def __get_last_day_with_name(year, month, day):
	start_day, length = monthrange(year, month)
	last_day = date(year, month, length)
	last_day_number = last_day.weekday()
	day_number = day_map[day]
	offset = (last_day_number - day_number + 7) % 7
	return last_day - timedelta(days=offset)

delta_map = {
	'1st': 0,
	'2nd': 7,
	'3rd': 14,
	'4th': 21,
	'5th': 28
}

day_map = {
	'Monday': 0,
	'Tuesday': 1,
	'Wednesday': 2,
	'Thursday': 3,
	'Friday': 4,
	'Saturday': 5,
	'Sunday': 6
}

File: javascript/meetup/meetup.js
const WEEK_START = {
    '1st': 1,
    '2nd': 8,
    '3rd': 15,
    '4th': 22,
    '5th': 29,
    'teenth': 13
};

const DAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

const getLastWeekStart = (year, month) => {
    return new Date(year, month + 1, 0).getDate() - 6;
};

module.exports = (year, month, day, descriptor) => {
    const weekStart = (descriptor === 'last') ? getLastWeekStart(year, month) : WEEK_START[descriptor];
    const weekStartDate = new Date(year, month, weekStart);
    const weekStartDay = weekStartDate.getDay();
    const dayNumber = DAYS.indexOf(day);
    const dayOffset = (dayNumber >= weekStartDay) ? dayNumber - weekStartDay : dayNumber + (7 - weekStartDay);
    const date = weekStartDate.getDate() + dayOffset;
    if (weekStartDate.getMonth() !== new Date(year, month).getMonth()) {
        throw 'Invalid date';
    }
    return new Date(year, month, date);
};

